scripts/submodule_updater.py

```diff
@@ -0,0 +1,174 @@
+        raise SystemExit(f'Repo {repo.working_dir} is missing remote \'{remote}\'')
+
+    url = getattr(repo.remotes, remote).url
+    pattern = r'^(git@|https://)([0-9a-zA-Z_.-]+)(:|/)([0-9a-zA-Z_.-]+)/([0-9a-zA-Z_-]+)(?:\.git)?'
+    match_remote = re.match(pattern, url)
+    if match_remote:
+        host = match_remote.group(2)
+        namespace = match_remote.group(4)
+        repo_name = match_remote.group(5)
+    else:
+        raise SystemExit(f'Failed to parse repo from {url}')
+
+    if host not in config_tokens:
+        raise SystemExit(f'API token for {host} not found')
+
+    if 'github' in host:
+        github_kwargs = {'login_or_token': config_tokens[host]}
+        if host != 'github.com':
+            github_kwargs['gh_url'] = 'https://' + host + '/api/v3'
+        gh = github.Github(**github_kwargs)
+        repo_api = gh.get_repo(f'{namespace}/{repo_name}')
+
+    else:
+        raise SystemExit(f'Host \'{host}\' not supported.')
+
+    return host, namespace, repo_name, repo_api
+
+if __name__ == "__main__":
+    # Parse arguments
+    args = parse_args()
+    head_remote = args.head_remote
+    head_branch = args.head_branch
+    base_remote = args.base_remote
+    base_branch = args.base_branch
+
+    # Load the config
+    with open(os.path.join(os.path.dirname(os.path.realpath(__file__)), 'config.yaml'), 'r') as f:
+        config = yaml.safe_load(f)
+
+    # Load the git repo
+    repo_path = os.path.realpath(args.repo)
+    try:
+        git_repo = git.Repo(repo_path)
+    except:
+        raise SystemExit(repo_path + ' is not a valid git repository')
+
+    # Load the remotes
+    head_host, head_namespace, head_repo, head_repo_api = load_remote(git_repo, head_remote, config['tokens'])
+    base_host, base_namespace, base_repo, base_repo_api = load_remote(git_repo, base_remote, config['tokens'])
+
+    # Sanity check on the same remotes
+    if head_host != base_host:
+       raise SystemExit(f'Host mismatch between remotes \'{head_remote}\' and \'{base_remote}\'')
+    host = head_host
+
+    # Load the user-provided submodules
+    submodule_sha1_args = vars(args)['submodule:sha1']
+    submodule_sha1 = {}
+    if submodule_sha1_args is not None:
+        for ss1 in submodule_sha1_args:
+            (sm_name, sha1) = ss1.split(':')
+            submodule_sha1[sm_name] = sha1
+
+    # Get list of active modules
+    submodules = []
+    for submodule in git_repo.submodules:
+        if submodule.module_exists():
+            submodules.append(submodule)
+
+    has_head_branch = head_branch in [branch.name for branch in git_repo.branches]
+    if has_head_branch:
+        print(f'Checking out branch \'{head_branch}\'')
+        git_repo.git.checkout(head_branch)
+    else:
+        print(f'Creating branch \'{head_branch}\'')
+        git_repo.git.checkout('-b', head_branch)
+
+    print(f'Resetting {head_branch} in {repo_path} to latest {base_remote}/{base_branch}')
+    git_repo.git.fetch(base_remote)
```

Comment:
what does this do on failed fetches?
I cant test I dont have the packages :S

Reply:
I tried specifying a remote that doesn't exist, but there is logic to catch that. I imagine that if the remote is valid and if there is a network issue during the fetch, some sort of crash will occur.